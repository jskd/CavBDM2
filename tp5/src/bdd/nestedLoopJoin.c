/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#include "nestedLoopJoin.h"
#include "bufferExtended.h"
#include "disk.h"

// See header
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out) {
  nested_loop_join(buf_a, buf_b, buf_out, NULL);
}

// See header
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out, FILE* overflow_file) {
  for(int a=0; a< buffer_count(buf_a); a++) {
    for(int b=0; b< buffer_count(buf_b); b++) {
      if(buffer_cmp(buf_a, a, buf_b, b) == 0) {
        // Overflow vidage dans overflow_file
        if(buffer_count(buf_out) == buffer_size(buf_out) && overflow_file != NULL) {
          buffer_write_file_from_descriptor(overflow_file, buf_out);
          buffer_flush(buf_out);
        }
        buffer_put_cpy(buf_out, buf_a, a);
        break; // pass to next a
      }
    }
  }
}

// See header
void nested_loop_join_disk( const struct disk* disk_a, struct buffer* buf_a,
    const struct disk* disk_b, struct buffer* buf_b,
    struct buffer* buf_out, FILE* overflow_file)
{
  char incr= 1;
  int b=0;
  int skip_reading=0;

  for(int a=0; a<disk_count(disk_a); a++) {
    buffer_read_file_from_descriptor( disk_item(disk_a, a),  buf_a);
    // Lecture en va-et-vient de disk b
    while( b < disk_count(disk_b) && b > -1 ) {
      // Saut de lecture si le contenu est déjà dans le buffer
      if(skip_reading)
        skip_reading= 0;
      else
        buffer_read_file_from_descriptor( disk_item(disk_b, b),  buf_b);
      nested_loop_join(buf_a, buf_b, buf_out, overflow_file);
      b+= incr;
    }
    incr= incr == 1 ? -1 : 1; // va-et-vient
    b+= incr; // annule le dépacement
    skip_reading= 1; // activate skip
  }
}
