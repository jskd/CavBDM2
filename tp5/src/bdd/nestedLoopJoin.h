/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#ifndef SRC_BDD_NESTEDLOOPJOIN_H
#define SRC_BDD_NESTEDLOOPJOIN_H

#include "bufferExtended.h"
#include "disk.h"

/**
 * Natual join (TP 1) Ne supporte pas les overflow
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du natural join de buf_a et buf_b
 */
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b,
    struct buffer* buf_out);

/**
 * nested_loop_join (TP5) Supporte les overflow (ecrit dans un fichier)
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du nested_loop_join de buf_a et buf_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b,
    struct buffer* buf_out, FILE* overflow_file);

/**
 * nested_loop_join sur disk (TP5)
 * Supporte les overflow (ecrit dans un fichier)
 *
 * @param[in]  disk_a  relation externe
 * @param[in]  buf_a   buffer de sauvegarde utilisé par le disk a
 * @param[in]  disk_b  relation interne
 * @param[in]  buf_b   buffer de sauvegarde utilisé par le disk b
 * @param[out] buf_out resultat du nested_loop_join du disk_a et disk_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void nested_loop_join_disk( const struct disk* disk_a, struct buffer* buf_a,
    const struct disk* disk_b, struct buffer* buf_b, struct buffer* buf_out,
    FILE* overflow_file);

#endif
