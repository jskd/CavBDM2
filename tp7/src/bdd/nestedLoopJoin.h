/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#ifndef SRC_BDD_NESTEDLOOPJOIN_H
#define SRC_BDD_NESTEDLOOPJOIN_H

#include "bufferExtended.h"
#include "diskReader.h"
#include "diskWriter.h"

/**
 * Natual join (TP 1) Ne supporte pas les overflow
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du natural join de buf_a et buf_b
 */
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b,
  struct buffer* buf_out);

/**
 * nested_loop_join (TP5) Supporte les overflow (ecrit dans un fichier)
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du nested_loop_join de buf_a et buf_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b,
  struct buffer* buf_out, struct disk_output* overflow_disk);

/**
 * nested_loop_join sur disk (TP5)
 * Supporte les overflow (ecrit dans un fichier)
 *
 * @param[in]  disk_a  relation externe
 * @param[in]  buf_a   buffer de sauvegarde utilisé par le disk a
 * @param[in]  disk_b  relation interne
 * @param[in]  buf_b   buffer de sauvegarde utilisé par le disk b
 * @param[out] buf_out resultat du nested_loop_join du disk_a et disk_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void nested_loop_join_disk( const struct diskReader* disk_a, struct buffer* buf_a,
  const struct diskReader* disk_b, struct buffer* buf_b, struct buffer* buf_out,
  struct disk_output* overflow_disk);

/**
 * table_bucket_join (TP6)
 * Fait la jointure de deux table
 *
 * @param[in]  tab_a   table a
 * @param[in]  buf_a   buffer de sauvegarde utilisé par la table a
 * @param[in]  tab_b   table b
 * @param[in]  buf_b   buffer de sauvegarde utilisé par le table b
 * @param[out] buf_out resultat du nested_loop_join du tab_a et tab_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void table_bucket_join(const struct table* tab_a, struct buffer* buf_a,
  const struct table* tab_b, struct buffer* buf_b,
  struct buffer* buf_out, struct disk_output* overflow_disk);

#endif
