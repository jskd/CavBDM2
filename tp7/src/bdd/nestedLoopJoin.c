/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#include "nestedLoopJoin.h"
#include "bufferExtended.h"
#include "table.h"
#include "diskReader.h"
#include "diskWriter.h"

// See header
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out) {
  nested_loop_join(buf_a, buf_b, buf_out, NULL);
}

// See header
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out, struct disk_output* overflow_disk) {
  for(int a=0; a< buffer_count(buf_a); a++) {
    for(int b=0; b< buffer_count(buf_b); b++) {
      if(buffer_cmp(buf_a, a, buf_b, b) == 0) {
        // Overflow vidage dans overflow_file
        if(buffer_isFull(buf_out) && overflow_disk != NULL) {
          buffer_write_file_from_descriptor( disk_output_get_current_file_descriptor(overflow_disk), buf_out);
          disk_output_next_file(overflow_disk);
          buffer_flush(buf_out);
        }
        buffer_put_cpy(buf_out, buf_a, a);
        break; // pass to next a
      }
    }
  }
}

// See header
void nested_loop_join_disk( const struct diskReader* disk_a, struct buffer* buf_a,
    const struct diskReader* disk_b, struct buffer* buf_b,
    struct buffer* buf_out, struct disk_output* overflow_disk)
{
  char incr= 1;
  int b=0;
  int skip_reading=0;

  for(int a=0; a<disk_r_count(disk_a); a++) {
    buffer_read_file_from_descriptor( disk_r_item(disk_a, a),  buf_a);
    // Lecture en va-et-vient de disk b
    while( b < disk_r_count(disk_b) && b > -1 ) {
      // Saut de lecture si le contenu est déjà dans le buffer
      if(skip_reading)
        skip_reading= 0;
      else
        buffer_read_file_from_descriptor( disk_r_item(disk_b, b),  buf_b);

      nested_loop_join(buf_a, buf_b, buf_out, overflow_disk);
      b+= incr;
    }
    incr= incr == 1 ? -1 : 1; // va-et-vient
    b+= incr; // annule le dépacement
    skip_reading= 1; // activate skip
  }
}

void table_bucket_join(const struct table* tab_a, struct buffer* buf_a,
  const struct table* tab_b, struct buffer* buf_b,
  struct buffer* buf_out, struct disk_output* overflow_disk) {

    if(table_get_n_bucket(tab_a) != table_get_n_bucket(tab_b)) {
      printf("Impossible de faire le table join sur des table n'ayant pas\
        le même nombre de bucket\n");
      return;
    }

    for(int bucket_index=0; bucket_index< table_get_n_bucket(tab_a); bucket_index++) {

       struct diskReader* disk_a= create_disk_from_bucket(tab_a, bucket_index);
       struct diskReader* disk_b= create_disk_from_bucket(tab_b, bucket_index);

       nested_loop_join_disk( disk_a, buf_a, disk_b, buf_b, buf_out, overflow_disk);

       disk_r_destroy(disk_a);
       disk_r_destroy(disk_b);
    }
}
