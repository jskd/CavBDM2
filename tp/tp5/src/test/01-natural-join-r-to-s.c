/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#include "../bdd/nestedLoopJoin.h"

static const size_t _buf_size= 10;
static const size_t _buf_data_lenght= 1;

int main(int argc, char** argv){


  // buffer R.txt
  struct buffer* buf_r= buffer_read_file("res/R.txt", _buf_size, _buf_data_lenght);
  if(buf_r == NULL) {
    printf("Erreur lors de la lecture de R.txt.\n");
    return -1;
  }

  // buffer S.txt
  struct buffer* buf_s= buffer_read_file("res/S.txt", _buf_size, _buf_data_lenght);
  if(buf_s == NULL) {
    printf("Erreur lors de la lecture de S.txt.\n");
    return -1;
  }

  // buffer de sortie
  struct buffer* buf_rs= buffer_create( _buf_size, _buf_data_lenght);

  // Jointure
  natural_join(buf_r, buf_s, buf_rs);


  buffer_dump(buf_rs);

  // Remove buf
  buffer_destroy(buf_r);
  buffer_destroy(buf_s);
  buffer_destroy(buf_rs);

  return 0;
}
