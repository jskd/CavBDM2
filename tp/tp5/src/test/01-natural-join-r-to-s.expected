buffer count: 6
buffer size: 10
buffer value::
  0000  5a 55 4b 45 42 56                                ZUKEBV
